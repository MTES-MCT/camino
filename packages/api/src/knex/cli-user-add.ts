import '../init.js'

// FIXME
const run = async () => {
  try {
    // await userAdd(knex, {
    //   id: 'admin',
    //   email: process.env.ADMIN_EMAIL,
    //   role: 'super',
    //   dateCreation: dateFormat(new Date(), 'yyyy-mm-dd')
    // })
  } catch (e) {
    console.error(e)
  } finally {
    process.exit(0)
  }
}

run()
