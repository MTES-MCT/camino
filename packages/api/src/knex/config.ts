import 'dotenv/config'
import { knexSnakeCaseMappers } from 'objection'
import path, { join } from 'node:path'
import { fileURLToPath } from 'url'
import { config } from '../config/index.js'
const __dirname = path.dirname(fileURLToPath(import.meta.url))

const connection = {
  host: config().PGHOST,
  port: config().PGPORT,
  database: config().PGDATABASE,
  user: config().PGUSER,
  password: config().PGPASSWORD,
}

export const simpleKnexConfig = {
  client: 'pg',
  connection,
  migrations: {
    directory: [join(__dirname, './migrations')],
    stub: join(__dirname, './migration-stub.ts'),
    extension: 'ts',
    loadExtensions: ['.ts'],
  },
}
export const knexConfig = {
  ...simpleKnexConfig,
  ...knexSnakeCaseMappers(),
}
