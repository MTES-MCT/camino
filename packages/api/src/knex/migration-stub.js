exports.up = async knex => {
  await knex.schema.dropTable('nexistepasEtVaFairePlanterLaMigration')

  return knex.schema.dropTable('nexistepasEtVaFairePlanterLaMigration')
}

exports.down = () => ({})
