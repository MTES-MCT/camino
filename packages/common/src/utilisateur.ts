import { Role, ROLES, utilisateurIdValidator } from './roles'
import { AdministrationId, IDS } from './static/administrations'
import { z } from 'zod'
import { entrepriseIdValidator } from './entreprise'

export const qgisTokenValidator = z.object({ token: z.string(), url: z.string() })
export type QGISToken = z.infer<typeof qgisTokenValidator>

export const utilisateurToEdit = z.object({
  id: utilisateurIdValidator,
  role: z.enum<Role, typeof ROLES>(ROLES),
  entreprises: z.array(entrepriseIdValidator),
  administrationId: z.enum<AdministrationId, typeof IDS>(IDS).nullable(),
})

export type UtilisateurToEdit = z.infer<typeof utilisateurToEdit>

export const newsletterAbonnementValidator = z.object({ newsletter: z.boolean() })

export const newsletterRegistrationValidator = z.object({
  email: z.string(),
})
