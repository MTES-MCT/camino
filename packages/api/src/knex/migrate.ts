import '../init'
import { knex } from '../knex'

const run = async () => {
  try {
    console.info('migrate…')

    const [latestBatchNo, latestLog] = await knex.migrate.latest()
    if (latestLog.length === 0) {
      console.info(`already up to date\n`)
    } else {
      console.info(`batch ${latestBatchNo} run: ${latestLog.length} migrations \n` + latestLog.join('\n') + '\n')
    }

    console.info('migrations terminées')
    process.exit()
  } catch (e) {
    if (e instanceof Error) {
      console.error(e.stack)
    } else {
      console.error(e)
    }

    process.exit(1)
  }
}

run()
