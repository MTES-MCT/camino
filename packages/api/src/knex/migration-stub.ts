import { Knex } from 'knex'

export const up = async (knex: Knex): Promise<void> => {
  await knex.schema.dropTable('nexistepasEtVaFairePlanterLaMigration')
}

export const down = (): void => {}
