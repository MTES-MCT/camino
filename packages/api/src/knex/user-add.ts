import { Knex } from 'knex'
import emailRegex from 'email-regex'
import bcrypt from 'bcryptjs'
import { IUtilisateur } from '../types'
import { idGenerate } from '../database/models/_format/id-create'

export const userAdd = async (
  knex: Knex,
  user: Omit<IUtilisateur, 'administrationId'>
): Promise<void> => {
  const password = idGenerate()
  const errors = []

  if (!user.email) {
    errors.push('email manquant')
  } else if (!emailRegex({ exact: true }).test(user.email)) {
    errors.push('adresse email invalide')
  }

  if (!errors.length) {
    user.motDePasse = bcrypt.hashSync(password, 10)

    await knex('utilisateurs').insert(user)

    console.info('Utilisateur créé')
    console.log(`utilisateur crée avec le mot de passe ${password}`)
  } else {
    console.info('Aucun user créé:', errors.join(', '))
  }
}
