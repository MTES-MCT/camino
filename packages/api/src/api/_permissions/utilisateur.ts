import { IUtilisateur, IUtilisateurCreation } from '../../types'
import { emailCheck } from '../../tools/email-check'
import { isAdministration, isEntreprise } from 'camino-common/src/roles'

const utilisateurEditionCheck = (
  utilisateur: IUtilisateur | IUtilisateurCreation
) => {
  const errors = []

  if (utilisateur.email && !emailCheck(utilisateur.email)) {
    errors.push('adresse email invalide')
  }

  if (
    !isAdministration(utilisateur) &&
    utilisateur.administrations &&
    utilisateur.administrations.length
  ) {
    errors.push(
      "le rôle de cet utilisateur ne permet pas de l'associer à une administration"
    )
  }

  if (
    !isEntreprise(utilisateur) &&
    utilisateur.entreprises &&
    utilisateur.entreprises.length
  ) {
    errors.push(
      "le rôle de cet utilisateur ne permet pas de l'associer à une entreprise"
    )
  }

  return errors
}

export { utilisateurEditionCheck }
