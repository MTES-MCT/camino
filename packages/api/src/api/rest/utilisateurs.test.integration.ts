import { restCall, restDeleteCall, restPostCall, userGenerate } from '../../../tests/_utils/index.js'
import { dbManager } from '../../../tests/db-manager.js'
import { Knex } from 'knex'
import { expect, test, describe, afterAll, beforeAll, vi } from 'vitest'
import { UtilisateurToEdit } from 'camino-common/src/utilisateur.js'
import type { Pool } from 'pg'
import { constants } from 'http2'

console.info = vi.fn()
console.error = vi.fn()
let knex: Knex<any, unknown[]>
let dbPool: Pool
beforeAll(async () => {
  const { knex: knexInstance, pool } = await dbManager.populateDb()
  dbPool = pool
  knex = knexInstance
})

afterAll(async () => {
  await dbManager.reseedDb()
  await dbManager.closeKnex()
})

describe('moi', () => {
  test('peut demander les informations sur soi-même', async () => {
    const user = await userGenerate({ role: 'defaut' })
    let tested = await restCall(dbPool, '/moi', {  }, undefined)
    
    expect(tested.statusCode).toBe(constants.HTTP_STATUS_NO_CONTENT)

    tested = await restCall(dbPool, '/moi', {  }, user)
    expect(tested.body).toMatchInlineSnapshot(`
      {
        "email": "defaut-user@camino.local",
        "id": "defaut-user",
        "nom": "nom-defaut",
        "prenom": "prenom-defaut",
        "role": "defaut",
      }
    `)
  })
})

describe('utilisateurModifier', () => {
  test('ne peut pas modifier un compte (utilisateur anonyme)', async () => {
    const utilisateurToEdit: UtilisateurToEdit = {
      id: 'test',
      role: 'defaut',
      entreprises: [],
      administrationId: null,
    }
    const tested = await restPostCall(dbPool, '/rest/utilisateurs/:id/permission', { id: utilisateurToEdit.id }, undefined, utilisateurToEdit)

    expect(tested.statusCode).toBe(403)
  })

  test("peut modifier le rôle d'un compte utilisateur", async () => {
    const userToEdit = await userGenerate({ role: 'defaut' })

    const utilisateurToEdit: UtilisateurToEdit = {
      id: userToEdit.id,
      role: 'admin',
      entreprises: [],
      administrationId: 'aut-97300-01',
    }
    const tested = await restPostCall(
      dbPool,
      '/rest/utilisateurs/:id/permission',
      { id: userToEdit.id },
      {
        role: 'super',
      },
      utilisateurToEdit
    )

    expect(tested.statusCode).toBe(204)
  })
})

describe('utilisateurSupprimer', () => {
  test('ne peut pas supprimer un compte (utilisateur anonyme)', async () => {
    const tested = await restDeleteCall(dbPool, '/rest/utilisateurs/:id', { id: 'test' }, undefined)
    expect(tested.statusCode).toBe(500)
    expect(tested.body).toMatchInlineSnapshot(`
      {
        "error": "aucun utilisateur avec cet id ou droits insuffisants pour voir cet utilisateur",
      }
    `)
  })

  test('peut supprimer son compte utilisateur', async () => {
    const user = await userGenerate({ role: 'defaut' })

    const tested = await restDeleteCall(dbPool, '/rest/utilisateurs/:id', { id: user.id }, { role: 'defaut' })
    expect(tested.statusCode).toBe(204)
  })

  test('peut supprimer un utilisateur (utilisateur super)', async () => {
    const id = 'user-todelete'
    await knex('utilisateurs').insert({
      id,
      prenom: 'userToDelete',
      nom: 'test',
      email: 'user-to-delete@camino.local',
      role: 'defaut',
      dateCreation: '2022-05-12',
    })

    const tested = await restDeleteCall(dbPool, '/rest/utilisateurs/:id', { id }, { role: 'super' })
    expect(tested.statusCode).toBe(204)
  })

  test('ne peut pas supprimer un utilisateur inexistant (utilisateur super)', async () => {
    const tested = await restDeleteCall(dbPool, '/rest/utilisateurs/:id', { id: 'not-existing' }, { role: 'super' })
    expect(tested.statusCode).toBe(500)
    expect(tested.body).toMatchInlineSnapshot(`
      {
        "error": "aucun utilisateur avec cet id ou droits insuffisants pour voir cet utilisateur",
      }
    `)
  })
})
