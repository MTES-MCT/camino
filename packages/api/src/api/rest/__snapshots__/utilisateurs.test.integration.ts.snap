// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getUtilisateurs > vérifie les droits de lecture > en tant que 'admin' 1`] = `
{
  "administrationId": "aut-97300-01",
  "email": "utilisateurEmail",
  "id": "utilisateurId",
  "nom": "utilisateurNom",
  "prenom": "prenom-pas-super",
  "role": "editeur",
  "telephoneFixe": null,
  "telephoneMobile": null,
}
`;

exports[`getUtilisateurs > vérifie les droits de lecture > en tant que 'defaut' 1`] = `
{
  "message": "droits insuffisants",
  "status": 403,
}
`;

exports[`getUtilisateurs > vérifie les droits de lecture > en tant que 'editeur' 1`] = `
{
  "administrationId": "aut-97300-01",
  "email": "utilisateurEmail",
  "id": "utilisateurId",
  "nom": "utilisateurNom",
  "prenom": "prenom-pas-super",
  "role": "editeur",
  "telephoneFixe": null,
  "telephoneMobile": null,
}
`;

exports[`getUtilisateurs > vérifie les droits de lecture > en tant que 'entreprise' 1`] = `
{
  "administrationId": "aut-97300-01",
  "email": "utilisateurEmail",
  "id": "utilisateurId",
  "nom": "utilisateurNom",
  "prenom": "prenom-pas-super",
  "role": "editeur",
  "telephoneFixe": null,
  "telephoneMobile": null,
}
`;

exports[`getUtilisateurs > vérifie les droits de lecture > en tant que 'lecteur' 1`] = `
{
  "administrationId": "aut-97300-01",
  "email": "utilisateurEmail",
  "id": "utilisateurId",
  "nom": "utilisateurNom",
  "prenom": "prenom-pas-super",
  "role": "editeur",
  "telephoneFixe": null,
  "telephoneMobile": null,
}
`;

exports[`getUtilisateurs > vérifie les droits de lecture > en tant que 'super' 1`] = `
{
  "administrationId": "aut-97300-01",
  "email": "utilisateurEmail",
  "id": "utilisateurId",
  "nom": "utilisateurNom",
  "prenom": "prenom-pas-super",
  "role": "editeur",
  "telephoneFixe": null,
  "telephoneMobile": null,
}
`;
